napari/utils/events/evented_model.py

```diff
@@ -75,3 +76,37 @@ def __setattr__(self, name, value):
     @property
     def events(self):
         return self._events
+
+    def asdict(self):
+        """Convert a model to a dictionary."""
+        warnings.warn(
+            (
+                "The `asdict` method has been renamed `dict` and is now deprecated. It will be"
+                " removed in 0.4.7"
+            ),
+            category=FutureWarning,
+            stacklevel=2,
+        )
+        return self.dict()
```

Comment:
I believe the issue here was more than just choice of wording.  pydantics `dict` recurses sub items in the model.  So if we want that behavior (which we do, for serialization) we have to use `dict` or do some deeper patching to the pydantic base model

Reply:
I see here a big problem with using these as serialization. This serialization method does not provide any information about field types, which makes it hard to write migration tool between versions.  

This is a nice idea when you would like to synchronize two applications in the same version, but when you would like to use it as long time storage, then it should be typed to use the Visitor approach.

Reply:
> This serialization method does not provide any information about field types, which makes it hard to write migration tool between versions.

actually, there is _fantastic_ builtin support for exactly this using the `object.schema()` and `object.schema_json()` methods (this is one of the things we like about pydantic).  It even supports nested schema (see `'$ref': '#/definitions/Address'` in this example).  For example:

```python
from typing import ClassVar
from napari.utils.events import EventedModel
from pydantic import BaseModel

class Address(BaseModel):
    number: int
    street: str

class User(EventedModel):
    """Demo evented model."""
    id: int
    name: str = 'Alex'
    address: Address
```

```python

>>> user = User(id=1, address={'number': '43', 'street': 'Main st.'})

>>> user
User(id=1, name='Alex', address=Address(number=43, street='Main st.'))

>>> user.dict()
{'id': 1, 'name': 'Alex', 'address': {'number': 43, 'street': 'Main st.'}}

>>> user.schema()
{'title': 'User',
 'description': 'Demo evented model.',
 'type': 'object',
 'properties': {'id': {'title': 'Id', 'type': 'integer'},
  'name': {'title': 'Name', 'default': 'Alex', 'type': 'string'},
  'address': {'$ref': '#/definitions/Address'}},
 'required': ['id', 'address'],
 'definitions': {'Address': {'title': 'Address',
   'type': 'object',
   'properties': {'number': {'title': 'Number', 'type': 'integer'},
    'street': {'title': 'Street', 'type': 'string'}},
   'required': ['number', 'street']}}}

>>> user.schema_json()
'{"title": "User", "description": "Demo evented model.", "type": "object", "properties": {"id": {"title": "Id", "type": "integer"}, "name": {"title": "Name", "default": "Alex", "type": "string"}, "address": {"$ref": "#/definitions/Address"}}, "required": ["id", "address"], "definitions": {"Address": {"title": "Address", "type": "object", "properties": {"number": {"title": "Number", "type": "integer"}, "street": {"title": "Street", "type": "string"}}, "required": ["number", "street"]}}}'
```


were you already aware of this and still concerned?

Reply:
> but stdlib uses asdict and for me, it is more natural.

also... that's just dataclasses... not a "global" standard lib thing.  These are no longer dataclasses so we saw it as less of a diversion from the stdlib syntax

Reply:
I see that it generate json schema, but not one structure with schema and values? 

You plan to dump schema and values every time?

Reply:
> You plan to dump schema and values every time?

@Czaki the serialization options with Pydantic are just much larger than what was available before, where recursion wasn't really working properly. I don't think we're really settling here exactly how we'll do serialization in each use case, but now we'll have more options

Reply:
As to the name itself, either we could leave both `dict` and `asdict` around (as an alias), but I don't love having aliases around and think it would be strange to try and remove the builtin `dict`.

Reply:
> You plan to dump schema and values every time?

erm, no... can you clarify the exact scenario you're worried about here?  I feel like there are more tools than usual in this arrangement for someone concerned about model serialzation and schema

> I see that it generate json schema, but not one structure with schema and values?

I've _rarely_ seen something serialize like that.  However, if you wanted to, we could add a _single_ `napari.__version__` item to each serialized object. So, if a developer were trying to deserialize some string from some _other_ version of napari, they could at least get the schema (is that the scenario you're talking about?  feels like we're getting way ahead of ourselves here... but even still, the pieces are in place)

Reply:
In the case of my software (PartSeg) I have a much slower change in core structures than napari, but I have persistence of state between sessions. This force to write migration schema. From version to version. 
For this, I use the `Visitor` pattern. 

I'm not sure if it could be simple to apply to napari. Or you have another plan for napari.

Reply:
We're discussing persistent settings on napari right now in other issues, but I don't think we've really discussed persistence of state between sessions. Maybe that's something you can open a new issue on @Czaki if you want to discuss further. I'd like to merge this PR now as i want to incorporate it's changes into another one already. I'll not that this PR doesn't actually use the EventedModel yet so until #2042 is merged it will have no effect

Reply:
But settings are part of the state. This is the same topic.

Reply:
Sorry I meant persistence of the entire state as opposed to parts of the state. We can continue the discussion around serialization of the settings on those dedicated PRs/ issues

Reply:
I also persist only part of the state and settings migrate will touch this code. But as it is merged then I agree that discussion should move to the next PR/issues.